### PR TITLE
#165273593 User sees read statistics

### PR DESCRIPTION
**Pull request by kalsmic**
[Finishes #165273593]
#### What does this PR do?
Adds reading stats to an article

#### Description of the task to be completed?
Add unit test for reading stats.
Add functionality for reading stats

#### Any background context you want to provide?
A reader can only add to the read stats on an article once hence viewing an article twice does not increase its view by two.
A reader can not add to the read stats of their own article.

#### How should this be manually tested?
View an article you have created and check on the statistics. They should not change
View an article created by another user and check on its statistics again.

#### What are the relevant pivotal tracker stories
[#165273593](https://www.pivotaltracker.com/story/show/165273593)

#### Postman screenshots
Get a specific article.
Screenshot 2019-05-06 at 22 02 34

### CODE REVIEW

**Review by bisonlou — CHANGES_REQUESTED** [review]
image

I get an error when am not authenticated, and try to get an article.

**Comment by kalsmic** [comment]
> image
> 
> I get an error when am not authenticated, and try to get an article.

Thanks, @bisonlou, let me fix this

File: authors/apps/articles/models.py [hunk 1]
```diff
@@ -28,3 +31,18 @@ class FavoriteArticle(models.Model):
...
+class ReadingStats(models.Model):
+    """
+    Model for Reading statistics
+    """
+
+    article = models.ForeignKey(Article, on_delete=models.CASCADE)
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+
+    class Meta:
+        unique_together = ('article', 'user')
```
Comment (bisonlou) [inline authors/apps/articles/models.py]: I see you have a unique together, which prohibits adding user and an article more than once. However I think a user should be able to read an article more than once, and each time the statistics should be captured.

**Review by bisonlou — CHANGES_REQUESTED** [review]

File: authors/apps/articles/views.py [hunk 2]
```diff
@@ -102,13 +85,27 @@ class UpdateDeleteArticleView(APIView):
...
+            try:
+
+                ReadingStats.objects.create(
+                    user=current_user,
+                    article=article
+                ).save()
+            except IntegrityError:
+                # Reading statistics for a given for per article are capture once for every reader
+                pass
+
```
Comment (AtamaZack) [inline authors/apps/articles/views.py]: Great work @kalsmic, awesome logic, however, given the same functionality on [medium.com](https://medium.com/), also anonymous users are allowed a read/view count.

File: authors/apps/articles/views.py [hunk 3]
```diff
@@ -102,13 +85,27 @@ class UpdateDeleteArticleView(APIView):
...
 
     def get(cls, request, article_id):
         article = get_object_or_404(Article, id=article_id, delete_status=False)
         serializer = ArticleSerializer(article, many=False)
         an_article = dict(serializer.data)
         an_article['likeCount'] = [like_grand_count(an_article)]
+
+        current_user = request.user
+        # Make this functionality on available for logged in users.
+        if current_user and article.author.username != current_user.username:
```
Comment (ZackAtama) [inline authors/apps/articles/views.py]: Great work @kalsmic, awesome logic, however, given the same functionality on [medium.com](https://medium.com/), also anonymous users are allowed a read/view count.

**Review by ZackAtama — CHANGES_REQUESTED** [review]

File: authors/apps/articles/models.py (hunk 1 above)
Comment (kalsmic) [reply authors/apps/articles/models.py]: @bison, I have removed the unique together validation and now a user can add to the view stats of the article each time they visit the article.

File: authors/apps/articles/views.py (hunk 3 above)
Comment (kalsmic) [reply authors/apps/articles/views.py]: @AtamaZack, I have implemented the feedback. now article views from an anonymous user add to the view stats of the article

**Review by bisonlou — APPROVED** [review]

**Review by ZackAtama — APPROVED** [review]

**Review by JamesMudidi — APPROVED** [review]
Nice work, the feature is working as expected.

**Review by llwasampijja — APPROVED** [review]
Nice work @kalsmic With the implementation of the changes requested by @bisonlou  and @AtamaZack , the work is ok for merging!

**Review by nadralia — APPROVED** [review]
Nice work

**Pull request merged**